JASP-Engine/JASP/R/principalcomponentanalysis.R

```diff
@@ -27,303 +28,384 @@
+  meta <- list(
+    list(name="title", type="title"),
+    list(name="factorLoadings", type = "table"),
+    list(name="factorCorrelations", type = "table"),
+    list(name="goodnessOfFit", type="table"),
+    list(name="fitMeasures", type="table"),
+    list(name="pathDiagram", type="image"),
+    list(name="screePlot", type="image"))
+
+  results[[".meta"]] = meta
+
+
+
+  ## initialize state ## ----
+
+  stateKey <- list(
+    analysisResults = c(
+      'rotationMethod',
+      'orthogonalSelector',
+      'obliqueSelector',
+      'variables', 
+      'factorMethod',
+      'eigenValuesBox',
+      'numberOfFactors'),
+    pathDiagram = c(
+      'rotationMethod',
+      'orthogonalSelector',
+      'obliqueSelector',
+      'variables', 
+      'factorMethod',
+      'eigenValuesBox',
+      'numberOfFactors',
+      'highlightSlider', 
+      'highlightText',
+      # 'incl_pathDiagram', 
+      'plotWidthPathDiagram',
+      'plotHeightPathDiagram'),
+    screePlot = c(
+      'rotationMethod',
+      'orthogonalSelector',
+      'obliqueSelector',
+      'variables', 
+      'factorMethod',
+      'eigenValuesBox',
+      'numberOfFactors',
+      # 'incl_screePlot',
+      'plotWidthScreePlot',
+      'plotHeightScreePlot')
+    )
+  keep <- NULL
+
+  # # States:
+  # newAnalysis <- TRUE
+  # state <- .retrieveState()
+
+  # if (!is.null(state) && !is.null(state[["analysisResults"]])) {  # is there state?
   
-  # States:
-  state <- .retrieveState()
-  analysisResults <- NULL
-  newAnalysis <- TRUE
-  # if ( ! is.null(state) && state$complete) {  # is there state?
-  
-  if ( !is.null(state)) {  # is there state?
-    
-    diff <- .diff(options, state$options)  # compare old and new options
+  # # if ( !is.null(state)) {  # is there state?
     
-    nVariable <- length(options$variables)
+  #   # nVariable <- length(options$variables)
+  #   diff <- .diff(options,state$options)
     
-    if (is.list(diff) && !diff[['rotationMethod']] && !diff[['orthogonalSelector']] && !diff[['obliqueSelector']] && !diff[['variables']] && !diff[['factorMethod']] && 
-        !diff[['eigenValuesBox']] && !diff[['numberOfFactors']]) {
-      
-      # old results can be used
-      newAnalysis <- FALSE
-      analysisResults <- state$analysisResults
-      
+  #   if (is.list(diff) && !diff[['rotationMethod']] && !diff[['orthogonalSelector']] && !diff[['obliqueSelector']] && !diff[['variables']] && !diff[['factorMethod']] && 
+  #       !diff[['eigenValuesBox']] && !diff[['numberOfFactors']]) {
+
+  #     # old results can be used
+  #     newAnalysis <- FALSE
+  #   }
+  # }
+
+  if(is.null(state[["analysisResults"]])){
+
+    state <- NULL
+    analysisResults <- NULL
+
+    if (perform == "run" && !is.null(dataset) && nrow(dataset) > 1){
+
+      error0 <- .hasErrors(dataset=dataset, perform=perform, type=c("infinity", "variance"), exitAnalysisIfErrors=TRUE)
+
+      customChecks <- list(
+        function() {
```

Comment:
@hummyhummy @TimKDJ custom checks in a common?

Reply:
now in the common main function called mainFunctionPCAEFA